Add debug info API for distributed autograd.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30642 Add debug info API for distributed autograd.**

Adding a couple of basic metrics for distributed autograd which would
help in determining stuckness.

Differential Revision: [D18776478](https://our.internmc.facebook.com/intern/diff/D18776478/)

Commit:
Add debug info API for distributed autograd.

Adding a couple of basic metrics for distributed autograd which would
help in determining stuckness.

Differential Revision: [D18776478](https://our.internmc.facebook.com/intern/diff/D18776478/)

[ghstack-poisoned]

satgera commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -280,6 +299,15 @@ size_t DistEngine::numBackwardPasses() const {
 
+std::unordered_map<std::string, std::string> DistEngine::getDebugInfo() const {
+  std::unordered_map<std::string, std::string> debugInfo;
+  debugInfo["num_current_backward_passes"] =
```
use constexpr instead of naked string.

satgera commented on torch/csrc/distributed/autograd/engine/dist_engine.h:
```diff
@@ -48,6 +48,10 @@ class TORCH_API DistEngine {
 
+  // Returns key-value pairs consisting of useful debugging information related
+  // to distributed autograd.
+  std::unordered_map<std::string, std::string> getDebugInfo() const;
```
This may qualify for getMetrics() instead of getDebugInfo().

pritamdamania87 replied:
Although this only has metrics currently, would like to keep this as debug info and str->str so we can easily put additional stuff in here as needed.

jjlilley commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -18,6 +18,24 @@ using torch::autograd::Node;
 
+namespace {
+
+static std::atomic<int> num_threads_blocked_in_backward = 0;
```
Should this instead be a member variable in DistEngine, especially because getDebugInfo() is a member method?

pritamdamania87 replied:
The guard is not a member of DistEngine and would need access to this variable and I'd like to keep the Guard definition in the cpp file.

mrshenli commented:
There will be conflicts between this one and #30610.

Commit:
Update on "Add debug info API for distributed autograd."


Adding a couple of basic metrics for distributed autograd which would
help in determining stuckness.

Differential Revision: [D18776478](https://our.internmc.facebook.com/intern/diff/D18776478/)

[ghstack-poisoned]

pritamdamania87 commented:
> There will be conflicts between this one and #30610.

I'll rebase on top of that once it lands.

mrshenli commented:
#30610 has just landed

Commit:
Update on "Add debug info API for distributed autograd."


Adding a couple of basic metrics for distributed autograd which would
help in determining stuckness.

Differential Revision: [D18776478](https://our.internmc.facebook.com/intern/diff/D18776478/)

[ghstack-poisoned]

Commit:
Update on "Add debug info API for distributed autograd."


Adding a couple of basic metrics for distributed autograd which would
help in determining stuckness.

Differential Revision: [D18776478](https://our.internmc.facebook.com/intern/diff/D18776478/)

[ghstack-poisoned]

Commit:
Update on "Add debug info API for distributed autograd."


Adding a couple of basic metrics for distributed autograd which would
help in determining stuckness.

Differential Revision: [D18776478](https://our.internmc.facebook.com/intern/diff/D18776478/)

[ghstack-poisoned]
